### ISSUE TITLE
Boot exception exposes database username and file system path

### ISSUE THREAD

**Issue by matteocontrini** [flarum/core#2290]
When Flarum cannot connect to the database it shows the following error message:

```
Flarum encountered a boot error (Doctrine\DBAL\Driver\PDOException)
SQLSTATE[HY000] [1045] Access denied for user 'flarum'@'localhost' (using password: YES)
thrown in /[REDACTED]/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOConnection.php on line 31
```

Therefore exposing the database username and the file system path to Flarum.

An improvement could be to print the exception only in debug mode, or even better to log the exception with the logger so that users never see it. Both cases require to split the boot phase in multiple calls.

Another proposal (by me) is to simply rethrow the exception after printing a generic error message, so that PHP takes care of it.

This was discussed more in detail with @clarkwinkelmann [on Discuss](https://discuss.flarum.org/d/24851-boot-exception-exposes-db-username-and-file-system-path).

**Comment by askvortsov1** [comment]
I think this could be expanded into a discussion of how to better handle boot errors in general, which is something I think would be quite nice

**Comment by awknode** [comment]
```
    private function safelyBootAndGetHandler()
    {
        try {
            return $this->site->bootApp()->getRequestHandler();
        } catch (Throwable $e) {
            error_log("The database was not available.", 3, "/var/tmp/my-errors.log");
            echo("Database is currently not available. See logs for more details.");
            exit(1);
        }
    }
```

this is in the core/Http/Server.php , i replaced their chunk of code with this and it works.

**Comment by askvortsov1** [comment]
Thank you to both of you for raising this and identifying a potential solution! A few immediate comments before we move to a PR stage:

- Instead of "Database is currently not available", I think "There was an error booting the site" would make more sense.
- I think we should write to the `storage/logs` folder instead of `/var/tmp`. I'd be fine just writing to a `boot_errors` file or something, but IMO it'd be nice to keep errors together. We should have access to 
- If in debug mode, it'd be nice to still show the formatted error, although the code needed to determine that might be too much.
- IIRC, doesn't Flarum's formatted error message only show when `display_errors` is enabled in PHP settings? If so, wouldn't it be better to add to the installation documentation that display_errors should be disabled in production? I might be wrong though.

**Comment by matteocontrini** [comment]
> IIRC, doesn't Flarum's formatted error message only show when `display_errors` is enabled in PHP settings? If so, wouldn't it be better to add to the installation documentation that display_errors should be disabled in production? I might be wrong though.

It seems to just spit the exception to the user with the `exit` function: https://github.com/flarum/core/blob/641619e8207c27e2649061f8f01754f6d1d45b3d/src/Http/Server.php#L53-L60

**Comment by awknode** [comment]
> > IIRC, doesn't Flarum's formatted error message only show when `display_errors` is enabled in PHP settings? If so, wouldn't it be better to add to the installation documentation that display_errors should be disabled in production? I might be wrong though.
> 
> It seems to just spit the exception to the user with the `exit` function:
> 
> https://github.com/flarum/core/blob/641619e8207c27e2649061f8f01754f6d1d45b3d/src/Http/Server.php#L53-L60

nope, i modified the php.ini file, php-fpm.d/www.conf file and the php-fpm.conf file in /etc/php7, none of this fixed the issue. I had to go into flarum itself.

**Comment by askvortsov1** [comment]
Looking into this a bit more, I think that using `exit` is likely problematic. Instead, perhaps we should throw a new exception with the formatted message? That way, display_errors should regulate whether it's logged or shown (and display_errors should NEVER be enabled in production).

> I personally love this project and that just pushed me over the edge as that password was one I used in another app so I had to make some adjustments after other users who use my site a lot notified me of what was going on. I almost got hacked and I know how to hack, which would have been shameful to me for pulling code and not having reviewed every bit of it before I ran it.

I agree that this is concerning: I don't think we've seen a case where the password itself was leaked, but now that we have, this is something that needs to be fixed.

**Comment by awknode** [comment]
> Looking into this a bit more, I think that using `exit` is likely problematic. Instead, perhaps we should throw a new exception with the formatted message? That way, display_errors should regulate whether it's logged or shown (and display_errors should NEVER be enabled in production).
> 
> > I personally love this project and that just pushed me over the edge as that password was one I used in another app so I had to make some adjustments after other users who use my site a lot notified me of what was going on. I almost got hacked and I know how to hack, which would have been shameful to me for pulling code and not having reviewed every bit of it before I ran it.
> 
> I agree that this is concerning: I don't think we've seen a case where the password itself was leaked, but now that we have, this is something that needs to be fixed.

https://paste.awknode.com/behujavipe.http
This is what i use in docker and i think it's because im linking 4 different apps to the same database, all it does is access that database but it transferred all the information with this setup and my obvious alterations here for docker.

**Comment by awknode** [comment]
```
<?php

// uncomment to see
// ini_set('display_errors', 0);

new PDO('a', 'b');
```
This is the actual fix if we want to still throw the exception and clear the important details out. Otherwise you can write to file like i've done and make a pretty page for it.

**Comment by clarkwinkelmann** [comment]
What about this?

Using `throw`, the error will only be visible with `display_errors=On`.

Using `echo`, we prevent a blank page when `display_errors=Off`.

```php
    private function safelyBootAndGetHandler()
    {
        try {
            return $this->site->bootApp()->getRequestHandler();
        } catch (Throwable $e) {
            echo 'Flarum encountered a boot error';

            throw $e;
        }
    }
```

By hiding the error message we make it a bit more complicated for forum owners to diagnose the problem, but it's probably not that bad.

It's not super common to have a boot error, and it's not too complicated to find the error in the system logs or temporarily enable `display_errors` to troubleshoot.

We might just need an additional note in our troubleshooting guide saying that if you see "Flarum encountered a boot error", then you should look in the system log files instead of Flarum log files.

**Comment by askvortsov1** [comment]
I like that! I think we could also say to look in the system log files inside that echo message to make things simpler.

**Comment by luceos** [comment]
What about this? (ready for pr)

```php
    private function safelyBootAndGetHandler()
    {
        try {
            return $this->site->bootApp()->getRequestHandler();
        } catch (Throwable $e) {

            if (app() && app()->has(LoggerInterface::class)) {
                app(LoggerInterface::class)->emergency(
                    $this->formatBootException($e)
                );
            }

            return new MaintenanceModeHandler;
        }
    }
    ```

**Comment by askvortsov1** [comment]
If there's a site boot error, it's 99% likely that it originates with app boot (extension or DB problem), not during the `getRequestHandler` part, so `app(LoggerInterface::class)` will almost never be available. 

I think I prefer Clark's approach of throwing an exception and echoing an error message (although I think that error message should include a note to look at the server's php/web server logs). It's simpler, and solves the issue when `display_errors` is off (as it should be on production).

**Comment by luceos** [comment]
I actually dislike echoing because the credentials are still put in buffer. If you don't like the logger implementation let's rely on error log which will send the error to the php error File.

**Comment by askvortsov1** [comment]
> credentials are still put in buffer

But the actual exception message wouldn't be displayed as long as display_errors is off though, I believe? We're echoing a generic "something has gone wrong check logs" error, not the message. Sorry perhaps I'm misunderstanding PHP's error/echo mechanism.

**Comment by luceos** [comment]
🤔 okay if we can confirm this solution works on both scenario's I'm fine with the proposition of @clarkwinkelmann

**Comment by clarkwinkelmann** [comment]
Further tests confirmed a way to reproduce the password exposure on my system. Inputting a wrong hostname like `localhost2` actually caused it on my system. For example:

`secret` is the database password

```
[2021-02-23 23:33:28] flarum.ERROR: PDOException: PDO::__construct(): php_network_getaddresses: getaddrinfo failed: Name or service not known in /home/clark/Projects/flarum-master/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOConnection.php:38
Stack trace:
#0 /home/clark/Projects/flarum-master/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOConnection.php(38): PDO->__construct('mysql:host=loca...', 'flarum-master', 'secret', Array)
#1 /home/clark/Projects/flarum-master/vendor/illuminate/database/Connectors/Connector.php(67): Doctrine\DBAL\Driver\PDOConnection->__construct('mysql:host=loca...', 'flarum-master', 'secret', Array)
#2 /home/clark/Projects/flarum-master/vendor/illuminate/database/Connectors/Connector.php(46): Illuminate\Database\Connectors\Connector->createPdoConnection('mysql:host=loca...', 'flarum-master', 'secret', Array)
#3 /home/clark/Projects/flarum-master/vendor/illuminate/database/Connectors/MySqlConnector.php(24): Illuminate\Database\Connectors\Connector->createConnection('mysql:host=loca...', Array, Array)
#4 /home/clark/Projects/flarum-master/vendor/illuminate/database/Connectors/ConnectionFactory.php(182): Illuminate\Database\Connectors\MySqlConnector->connect(Array)
#5 [internal function]: Illuminate\Database\Connectors\ConnectionFactory->Illuminate\Database\Connectors\{closure}()
#6 /home/clark/Projects/flarum-master/vendor/illuminate/database/Connection.php(924): call_user_func(Object(Closure))
#7 /home/clark/Projects/flarum-master/vendor/illuminate/database/Connection.php(906): Illuminate\Database\Connection->getPdo()
#8 /home/clark/Projects/flarum-master/vendor/illuminate/database/Connection.php(892): Illuminate\Database\Connection->getDoctrineConnection()
#9 /home/clark/Projects/flarum-master/workbench/core/src/Database/Migrator.php(67): Illuminate\Database\Connection->getDoctrineSchemaManager()
#10 [internal function]: Flarum\Database\Migrator->__construct(Object(Flarum\Database\DatabaseMigrationRepository), Object(Illuminate\Database\MySqlConnection), Object(Illuminate\Filesystem\Filesystem))
#11 /home/clark/Projects/flarum-master/vendor/illuminate/container/Container.php(843): ReflectionClass->newInstanceArgs(Array)
#12 /home/clark/Projects/flarum-master/vendor/illuminate/container/Container.php(681): Illuminate\Container\Container->build('Flarum\\Database...')
#13 /home/clark/Projects/flarum-master/vendor/illuminate/container/Container.php(629): Illuminate\Container\Container->resolve('Flarum\\Database...', Array)
#14 /home/clark/Projects/flarum-master/vendor/illuminate/container/Container.php(945): Illuminate\Container\Container->make('Flarum\\Database...')
#15 /home/clark/Projects/flarum-master/vendor/illuminate/container/Container.php(873): Illuminate\Container\Container->resolveClass(Object(ReflectionParameter))
#16 /home/clark/Projects/flarum-master/vendor/illuminate/container/Container.php(834): Illuminate\Container\Container->resolveDependencies(Array)
#17 /home/clark/Projects/flarum-master/vendor/illuminate/container/Container.php(261): Illuminate\Container\Container->build('Flarum\\Extensio...')
#18 /home/clark/Projects/flarum-master/vendor/illuminate/container/Container.php(799): Illuminate\Container\Container->Illuminate\Container\{closure}(Object(Illuminate\Container\Container), Array)
#19 /home/clark/Projects/flarum-master/vendor/illuminate/container/Container.php(681): Illuminate\Container\Container->build(Object(Closure))
#20 /home/clark/Projects/flarum-master/vendor/illuminate/container/Container.php(629): Illuminate\Container\Container->resolve('Flarum\\Extensio...', Array)
#21 /home/clark/Projects/flarum-master/workbench/core/src/Extension/ExtensionServiceProvider.php(30): Illuminate\Container\Container->make('flarum.extensio...')
#22 [internal function]: Flarum\Extension\ExtensionServiceProvider->Flarum\Extension\{closure}(Object(Flarum\Foundation\Application))
#23 /home/clark/Projects/flarum-master/workbench/core/src/Foundation/Application.php(322): call_user_func(Object(Closure), Object(Flarum\Foundation\Application))
#24 /home/clark/Projects/flarum-master/workbench/core/src/Foundation/Application.php(263): Flarum\Foundation\Application->fireAppCallbacks(Array)
#25 /home/clark/Projects/flarum-master/workbench/core/src/Foundation/InstalledSite.php(148): Flarum\Foundation\Application->boot()
#26 /home/clark/Projects/flarum-master/workbench/core/src/Foundation/InstalledSite.php(81): Flarum\Foundation\InstalledSite->bootLaravel()
#27 /home/clark/Projects/flarum-master/workbench/core/src/Http/Server.php(58): Flarum\Foundation\InstalledSite->bootApp()
#28 /home/clark/Projects/flarum-master/workbench/core/src/Http/Server.php(35): Flarum\Http\Server->safelyBootAndGetHandler()
#29 /home/clark/Projects/flarum-master/public/index.php(22): Flarum\Http\Server->listen()
#30 {main}
```

**Comment by clarkwinkelmann** [comment]
Some additional things I have decided based on my research:

Most errors, including database related errors, happen after the config and logger have been initialized, this means the vast majority of boot errors will happen with those components being available.

So I've decided that if the config is accessible, we will look for debug mode, and if found, use the existing behavior.

Otherwise we will check if the logger is ready, and if so, log the error to our own log file. This will be a lot more convenient to help forum owners. The message is modified to say that the error will be found in the Flarum log file.

If neither debug mode nor the logger is available, we default to my proposed strategy of echoing a generic message and throwing the error. This generic message has been updated to mention specifically "system PHP log file".

I also noticed we were not throwing a 500 error code. I have changed that in the PR
